File: leetcode/easy/283-moveZeroes.java
// first pass solution
class Solution {
    public void moveZeroes(int[] nums) {
        int left = 0;
        
        for (int num : nums) {
            if (num != 0) {
                nums[left] = num;
                left++;
            }
        }
        while (left < nums.length) {
            nums[left] = 0;
            left++;
        }
    }
}

File: leetcode/easy/283-moveZeroes.cpp
// first pass 
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int left = 0;
        int right = 0;

        while (right < nums.size()) {
            if (nums[right] == 0) {
                while (right < nums.size() && nums[right] == 0) right++;
            }
            if (right < nums.size() && nums[right] != 0) swap(nums[left], nums[right]);
            right++;
            left++;
        }
    }
};

// second pass
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int left = 0;
        int right = 0;

        while (right < nums.size()) {
            if (nums[left] == 0) swap(nums[right], nums[left]);
            if (nums[left] != 0) left++;
            right++;
        }
    }
};
